ocaml-lsp-server/src/dune.ml

```diff
@@ -1,46 +1,159 @@
 open Import
 
+module Chan = struct
+  open Fiber.O
+
+  type t =
+    { in_thread : Scheduler.thread
+    ; out_thread : Scheduler.thread
+    ; out_chan : out_channel
+    ; in_chan : in_channel
+    }
+
+  let write t sexp =
+    let+ res =
+      Scheduler.async_exn t.out_thread (fun () ->
+          match sexp with
+          | None -> close_out_noerr t.out_chan
+          | Some sexp -> Csexp.to_channel t.out_chan sexp)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let read t =
+    let+ res =
+      Scheduler.async_exn t.in_thread (fun () ->
+          match Csexp.input_opt t.in_chan with
+          | Error _
+          | Ok None ->
+            Unix.close (Unix.descr_of_in_channel t.in_chan);
+            None
+          | Ok (Some s) -> Some s)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let create scheduler in_fd out_fd =
+    let in_thread = Scheduler.create_thread scheduler in
+    let out_thread = Scheduler.create_thread scheduler in
+    let in_chan = Unix.in_channel_of_descr in_fd in
+    let out_chan = Unix.out_channel_of_descr out_fd in
+    { in_thread; out_thread; in_chan; out_chan }
+end
+
+module Drpc = Dune_rpc.V1
+
+module Client =
+  Drpc.Client
+    (struct
```

Comment:
Oh, `Drpc.Client` is functorized? I thought it depended on `stdune` already, why not use Dune's `Fiber`?

Reply:
Two reasons:

* `Fiber` isn't public yet.
* We want people to write clients with Lwt, Async, and other concurrency libraries they might think of.